### PR TITLE
trust signer-add

### PR DESCRIPTION
**Pull request by ashfall**
Also moved some helper functions I am reusing from `sign.go` to `helpers.go`.

Example use:
```
./build/docker-darwin-amd64 trust signer-add zed ashfall/test ashfall/test-demo --key ~/demo-keys/cert.pub

Adding signer "zed" to ashfall/test...
Enter passphrase for repository key with ID b5ed45c:
Successfully added signer: zed to ashfall/test

Adding signer "zed" to ashfall/test-demo...
Enter passphrase for repository key with ID 1500c31:
Successfully added signer: zed to ashfall/test-demo
```

image

### CODE REVIEW

File: cli/command/trust/signer_add.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+	importers, err := getImporters(trustDir)
+	if err != nil {
+		return err
+	}
+	if err := importKeys(newSignerRoleName.String(), options.key, passRetriever, importers); err != nil {
+		return err
+	}
+
+	// Now retrieve the notary key to add signer
+	signerKey, err := getOrGenerateNotaryKey(notaryRepo, newSignerRoleName)
```
Comment (ashfall) [inline cli/command/trust/signer_add.go]: @riyazdf Wondering if there's a way to eliminate this call altogether and have utils.ImportKeys give us the key as it adds?

File: cli/command/trust/signer_add.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+)
+
+type signerAddOptions struct {
+	key string
+}
+
+func newSignerAddCommand(dockerCli command.Cli) *cobra.Command {
+	options := signerAddOptions{}
+	cmd := &cobra.Command{
+		Use:   "signer-add [OPTIONS] IMAGE NAME",
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: can we switch this to `NAME IMAGE [IMAGE...]`?   Alternatively, I could see an argument for having images specified by (potentially multiple) `--repo` flags

File: cli/command/trust/signer_add.go [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+	if options.key == "" {
+		return fmt.Errorf("No key provided. Please use the --key flag to provide the path to a valid key")
+	}
+
+	_, ref, repoInfo, authConfig, err := getImageReferencesAndAuth(cli, imageName)
+	if err != nil {
+		return err
+	}
+
+	// TODO: initialize the repo if it's not already initialized with notary?
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: we can probably refactor the code from `sign.go` that handles initialization to a helper to be shared here

File: cli/command/trust/signer_add.go (hunk 1 above)
Comment (riyazdf) [reply cli/command/trust/signer_add.go]: yeah we could probably have `importKeys` return a list of keyIDs that it successfully imported - I think that would be a fine addition cc @endophage

File: cli/command/trust/signer_add.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	notaryRepo, err := trust.GetNotaryRepository(cli, repoInfo, *authConfig, "push", "pull")
+	if err != nil {
+		return trust.NotaryError(ref.Name(), err)
+	}
+
+	newSignerRoleName := data.RoleName(path.Join(data.CanonicalTargetsRole.String(), signerName))
+
+	passRetriever := passphrase.PromptRetrieverWithInOut(cli.In(), cli.Out(), nil)
+	trustDir := trust.GetTrustDirectory()
+	importers, err := getImporters(trustDir)
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: super nit: we can collapse the above line into this one

File: cli/trust/trust.go [hunk 5]
```diff
@@ -36,7 +36,8 @@ var (
 	ReleasesRole = data.RoleName(path.Join(data.CanonicalTargetsRole.String(), "releases"))
 )
 
-func trustDirectory() string {
+// GetTrustDirectory returns the bwase trust directory name
```
Comment (riyazdf) [inline cli/trust/trust.go]: typo: base

**Review by riyazdf — COMMENTED** [review]
this is coming along nicely - would be good to get multi-repo support in as well 👍

File: cli/command/trust/signer_add.go (hunk 2 above)
Comment (ashfall) [reply cli/command/trust/signer_add.go]: Hm, reordered this too `IMAGE NAME` for starters. As I add multi-repo support, liking the idea of `--repo` flag more.

`signer-add alice --repo alpine --repo linuxkit/alpine --key /path/to/key` looks nicer to me. Thoughts?

Comment (riyazdf) [reply cli/command/trust/signer_add.go]: 👍  looks good!

File: cli/command/trust/signer_add.go [hunk 6]
```diff
@@ -0,0 +1,129 @@
...
+type signerAddOptions struct {
+	key    string
+	signer string
+	images opts.ListOpts
+}
+
+func newSignerAddCommand(dockerCli command.Cli) *cobra.Command {
+	var options signerAddOptions
+	cmd := &cobra.Command{
+		Use:   "signer-add [OPTIONS] NAME ",
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: from feedback: `signer-add [OPTIONS] NAME IMAGE [IMAGE...]`

File: cli/command/trust/signer_add.go [hunk 7]
```diff
@@ -0,0 +1,129 @@
...
+		Short: "Add a signer",
+		Args:  cli.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.signer = args[0]
+			return addSigner(dockerCli, &options)
+		},
+	}
+	flags := cmd.Flags()
+	flags.StringVarP(&options.key, "key", "k", "", "Path to the signer key to add")
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: from feedback: this can become a `opts.NewListOpts`

File: cli/command/trust/signer_add.go [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+	fmt.Fprintf(cli.Out(), "\nAdding signer \"%s\" to %s...\n", signerName, imageName)
+	newSignerRoleName := data.RoleName(path.Join(data.CanonicalTargetsRole.String(), signerName))
+
+	passRetriever := passphrase.PromptRetrieverWithInOut(cli.In(), cli.Out(), nil)
+	importers, err := getImporters(trust.GetTrustDirectory())
+	if err != nil {
+		return err
+	}
+	if err := importKeys(newSignerRoleName.String(), keyPath, passRetriever, importers); err != nil {
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: from discussion: this might need to take `signerName` instead of the full `targets/<signerName>`

File: cli/command/trust/sign.go [hunk 9]
```diff
@@ -148,14 +148,7 @@ func prettyPrintExistingSignatureInfo(cli command.Cli, existingSigInfo trustTagR
...
 func initNotaryRepoWithSigners(notaryRepo *client.NotaryRepository, newSigner data.RoleName) error {
-	rootKey, err := getOrGenerateNotaryKey(notaryRepo, data.CanonicalRootRole)
-	if err != nil {
-		return err
-	}
-	rootKeyID := rootKey.ID()
-
-	// Initialize the notary repository with a remotely managed snapshot key
-	if err := notaryRepo.Initialize([]string{rootKeyID}, data.CanonicalSnapshotRole); err != nil {
+	if err := getOrGenerateRootKeyAndInitRepo(notaryRepo); err != nil {
```
Comment (riyazdf) [inline cli/command/trust/sign.go:151]: 👍  thanks for refactoring this

File: cli/command/trust/sign_test.go [hunk 10]
```diff
@@ -289,3 +161,15 @@ func TestChangeList(t *testing.T) {
 	cl, err := notaryRepo.GetChangelist()
 	assert.Equal(t, len(cl.List()), 0)
 }
+
+func TestInitNotaryRepoWithSigners(t *testing.T) {
```
Comment (riyazdf) [inline cli/command/trust/sign_test.go]: can we change this test name to reflect the offline nature for now?  `TestInitOfflineNotaryRepoWithSigners` or similar?

File: cli/command/trust/signer_add.go [hunk 11]
```diff
@@ -0,0 +1,115 @@
...
+		Args:  cli.RequiresMinArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.signer = args[0]
+			options.images = args[1:]
+			return addSigner(dockerCli, &options)
+		},
+	}
+	flags := cmd.Flags()
+	options.keys = opts.NewListOpts(nil)
+	flags.VarP(&options.keys, "key", "k", "Path to the signer key to add")
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: can we mention that this is a public key?

File: cli/command/trust/signer_add.go [hunk 12]
```diff
@@ -0,0 +1,115 @@
...
+	return cmd
+}
+
+func addSigner(cli command.Cli, options *signerAddOptions) error {
+	signerName := options.signer
+
+	// TODO@ashfall: instead of failing if no key is provided, ask user if
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: can we open a tracking issue for this TODO?

File: cli/command/trust/signer_add.go [hunk 13]
```diff
@@ -0,0 +1,115 @@
...
+	// they want to generate a key
+	if options.keys.Len() < 1 {
+		return fmt.Errorf("path to a valid key must be provided")
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: should note public key here as well

File: cli/command/trust/signer_add.go [hunk 14]
```diff
@@ -0,0 +1,115 @@
...
+
+	notaryRepo, err := trust.GetNotaryRepository(cli, repoInfo, *authConfig, "push", "pull")
+	if err != nil {
+		return trust.NotaryError(ref.Name(), err)
+	}
+
+	if err = notaryRepo.Update(false); err != nil {
+		switch err.(type) {
+		case client.ErrRepoNotInitialized, client.ErrRepositoryNotExist:
+			fmt.Fprintf(cli.Out(), "Initializing %s...\n", imageName)
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: nit: maybe this should make a mention of the signed repository: `Initializing signed repository for %s...`?

File: cli/command/trust/signer_add.go [hunk 15]
```diff
@@ -0,0 +1,115 @@
...
+			if err := getOrGenerateRootKeyAndInitRepo(notaryRepo); err != nil {
+				return trust.NotaryError(ref.Name(), err)
+			}
+			fmt.Fprintf(cli.Out(), "Successfully initialized %q\n", notaryRepo.GetGUN().String())
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: let's use image name instead of GUN for consistency.  The similar block in `sign.go` might need to be updated as well, can be a separate PR

File: cli/command/trust/signer_add.go [hunk 16]
```diff
@@ -0,0 +1,115 @@
...
+		},
+	}
+	flags := cmd.Flags()
+	options.keys = opts.NewListOpts(nil)
+	flags.VarP(&options.keys, "key", "k", "Path to the signer key to add")
+	return cmd
+}
+
+func addSigner(cli command.Cli, options *signerAddOptions) error {
+	signerName := options.signer
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go:44]: do we want to make a special case for `signerName == releases`?  I suppose this is a very rare case, though

File: cli/command/trust/signer_add_test.go [hunk 17]
```diff
@@ -0,0 +1,112 @@
...
+	for _, tc := range testCases {
+		cmd := newSignerAddCommand(
+			test.NewFakeCli(&fakeClient{}))
+		cmd.SetArgs(tc.args)
+		cmd.SetOutput(ioutil.Discard)
+		testutil.ErrorContains(t, cmd.Execute(), tc.expectedError)
+	}
+}
+
+func TestSignerAddCommand(t *testing.T) {
```
Comment (riyazdf) [inline cli/command/trust/signer_add_test.go]: nit: can we rename this test to `TestSignerAddCommandNoTargetsKey`?

**Review by riyazdf — COMMENTED** [review]
this is looking really good, mostly cosmetic feedback :+1:

**Comment by riyazdf** [comment]
also could you please squash on the next commit?  It will make the upstream PR and rebasing simpler :)

File: cli/command/trust/signer_add.go (hunk 16 above)
Comment (ashfall) [reply cli/command/trust/signer_add.go:44]: Yeah, rare case, but it leads to a replacement of the key 🔢 . Updating, thanks!

File: cli/command/trust/signer_add.go (hunk 12 above)
Comment (ashfall) [reply cli/command/trust/signer_add.go]: Sounds good, killed the TODO, and filed https://github.com/riyazdf/cli/issues/90

File: cli/command/trust/signer_add.go [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+			fmt.Fprintf(cli.Out(), "Successfully initialized %q\n", imageName)
+		default:
+			return trust.NotaryError(repoInfo.Name.Name(), err)
+		}
+	}
+
+	fmt.Fprintf(cli.Out(), "\nAdding signer \"%s\" to %s...\n", signerName, imageName)
+	newSignerRoleName := data.RoleName(path.Join(data.CanonicalTargetsRole.String(), signerName))
+
+	signerPubKeys, err := ingestPublicKeys(keyPaths)
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go:91]: Can we handle this `err` and error out if it's not nil?

File: cli/command/trust/signer_add.go [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+		Args:  cli.RequiresMinArgs(2),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			options.signer = args[0]
+			options.images = args[1:]
+			return addSigner(dockerCli, &options)
+		},
+	}
+	flags := cmd.Flags()
+	options.keys = opts.NewListOpts(nil)
+	flags.VarP(&options.keys, "key", "k", "Path to the signer's public key")
```
Comment (riyazdf) [inline cli/command/trust/signer_add.go]: nit: let's change `key` to `key(s)` to indicate that the signer could have multiple keys

**Review by riyazdf — APPROVED** [review]
LGTM on green
